network/openssl_privatekey.py

```diff
@@ -0,0 +1,281 @@
+        description:
+            - Name of the folder in which the generated SSL private key will be written
+    publickey_path:
+        required: false
+        description:
+            - Name of the folder in which the generated SSL public key will be written
+'''
+
+EXAMPLES = '''
+# Generate an OpenSSL private key with the default values (4096 bits, RSA)
+# and no public key
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem
+
+# Generate an OpenSSL private key with the default values (4096 bits, RSA)
+# and its matching public key (Works only when PyOpenSSL > 16.0.0)
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem
+                      publickey_path=/etc/ssl/public/ansible.com.pem
+
+# Generate an OpenSSL private key with a different size (2048 bits)
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem size=2048
+
+# Force regenerate an OpenSSL private key if it already exists
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem force=True
+
+# Generate an OpenSSL private key with a different algorithm (DSA)
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem type=DSA
+'''
+
+RETURN = '''
+size:
+    description: Size (in bits) of the SSL private key
+    returned:
+        - changed
+        - success
+    type: integer
+    sample: 4096
+type:
+    description: Algorithm used to generate the SSL private key
+    returned:
+        - changed
+        - success
+    type: string
+    sample: RSA
+filename:
+    description: Path to the generated SSL private key file
+    returned:
+        - changed
+        - success
+    type: string
+    sample: /etc/ssl/private/ansible.com.pem
+publickey_filename:
+    description: Path to the matching SSL public key file
+    returned:
+        - changed
+        - success
+    type: string
+    sample: /etc/ssl/public/ansible.com.pem
+'''
+
+class PrivateKeyError(Exception):
+    pass
+
+class PrivateKey(object):
+
+    def __init__(self, module):
+        self.size = module.params['size']
+        self.state = module.params['state']
+        self.name = os.path.basename(module.params['path'])
+        self.type = module.params['type']
+        self.force = module.params['force']
+        self.path = module.params['path']
+        self.changed = True
+        self.check_mode = module.check_mode
+        self.publickey_path = module.params['publickey_path']
+
+    def generate(self):
+        """Generate a keypair."""
+
+        if not self.force and os.path.exists(self.path):
+            self.changed = False
+            return
+
+        self.privatekey = crypto.PKey()
+
+        if self.type == 'RSA':
+            crypto_type = crypto.TYPE_RSA
+        else:
+            crypto_type = crypto.TYPE_DSA
+
+        try:
+            self.privatekey.generate_key(crypto_type, self.size)
+        except (TypeError, ValueError):
+            raise PrivateKeyError(get_exception())
+
+        try:
+            privatekey_file = open(self.path, 'w')
```

Comment:
Whats the default file mode end up here? 

Should probably be fairly restricted.

Maybe this module should use the default file args arg spec as well?  (for path/mode/etc)